pkg/compiler/analysis.go

```diff
@@ -138,26 +142,48 @@ func (c *codegen) traverseGlobals() bool {
 // countGlobals counts the global variables in the program to add
 // them with the stack size of the function.
 // Second returned argument contains the amount of global constants.
-func countGlobals(f ast.Node) (int, int) {
-	var numVar, numConst int
+func countGlobals(f ast.Node) (int, int, int) {
+	var numVar, numUnusedVarWithFunc, numConst int
 	ast.Inspect(f, func(node ast.Node) bool {
 		switch n := node.(type) {
 		// Skip all function declarations if we have already encountered `defer`.
 		case *ast.FuncDecl:
 			return false
 		// After skipping all funcDecls, we are sure that each value spec
-		// is a global declared variable or constant.
+		// is a globally declared variable or constant.
 		case *ast.GenDecl:
 			isVar := n.Tok == token.VAR
 			if isVar || n.Tok == token.CONST {
 				for _, s := range n.Specs {
-					for _, id := range s.(*ast.ValueSpec).Names {
+					valueSpec := s.(*ast.ValueSpec)
+					multiRet := len(valueSpec.Values) != 0 && len(valueSpec.Names) != len(valueSpec.Values) // e.g. var A, B = f() where func f() (int, int)
+					for j, id := range s.(*ast.ValueSpec).Names {
 						if id.Name != "_" {
 							if isVar {
 								numVar++
 							} else {
 								numConst++
 							}
+						} else {
+							if isVar && len(valueSpec.Values) != 0 {
+								var hasFuncCall bool
+								if j == 0 || !multiRet {
+									// Search for function calls inside the variable declaration.
+									ast.Inspect(valueSpec.Values[j], func(node ast.Node) bool {
+										switch node.(type) {
+										case *ast.CallExpr:
+											hasFuncCall = true
+										case *ast.Ident:
+											// Can safely skip idents immediately, we're interested at function calls only.
```

Comment:
`var _ = A + f()`?

Reply:
This case will be properly handled. Return value for `ast.Inspect`'s `f` function is whether to continue *in-depth* traversal. Here we don't need to go deeper into `A` (which doesn't have children anyway, but we can save at least `f(nil)` call). instead, we'll start from the next child which is f().

Reply:
Am I missing something?

Reply:
> in-depth

Yep, that solves it.